src/architecture/services.md

```diff
@@ -242,28 +240,33 @@ Transactions are executed during [the precommit stage](consensus.md)
 of the consensus (this concerns validators only) or when a node receives a
 block.
 
-### Event handling
+### Event Handling
 
 Services may subscribe to events (such as a block being committed) and perform
-some work in the event handler. The event handlers cannot modify the blockchain
-state, but can be used for various tasks such as logging, data migrations,
+some work in the event handlers. Some event handlers cannot modify the
+blockchain state, while some handlers can. See
+[Commit Handler](#commit-handler) section for more information on the applied
+handlers.
+
+The handlers can be used for various tasks such as logging, data migrations,
 updating local parameters, and/or generating and broadcasting transactions to
 the blockchain network.
 
 !!! note
-    As of Exonum 0.1, the only built-in event is block commit. More events
+    Currently, the only built-in event for services subscription is block
+    commit. The handlers operable with this event are `before_commit` and
+    `after_commit`. More events
     will be added in the future, including possibility for services to define
     and emit events and for services and light clients to subscribe to events
     emitted by the services.
 
 ## Service Development
 
 !!! note
-    As of Exonum 0.1, you can only code services in
-    [Rust](http://rust-lang.org/).
-    Rust is probably the safest general-purpose programming language, but it is
-    not very easy to master. Java binding
-    [is a high-priority task](../roadmap.md).
+    You can code Exonum services in [Rust](http://rust-lang.org/) or Java.
+    Rust has been chosen as probably the safest general-purpose programming
+    language, but it is not very easy to master. To develop Exonum services in
+    Java, use the [Java Binding tool][java-binding].
```

Comment:
Shall we reference the get-started/java-binding instead? Otherwise, the interested reader would go to repo, then to ejb module, just to see they shall go back to docs.

Reply:
done